Code Review

thgoebel opened the pull request:
@felixlinker

thgoebel committed:
Add first model outline

thgoebel committed:
Fix syntax errors + wellformedness checks

thgoebel committed:
WIP: resolve some partial decs left, refactor server persistent facts as SST_

thgoebel committed:
Reformat action facts for readability

The mostly-white lines that now separate premise from action facts,
and action facts from conclusion make it easier to read.
The three things (premise, actions, conclusion) are not so clumped together.

thgoebel committed:
Add midway presentation slides

thgoebel committed:
Fix strictly-less-than vs. less-or-equal-than

thgoebel committed:
Add license

thgoebel committed:
Work around PDs in insertion msg

thgoebel committed:
Give server an sk

thgoebel committed:
Server: sign published epochs

WIP: the signatures are not yet used anywhere

thgoebel committed:
Server: only store latest_epoch_id + move other epoch data into SST_

thgoebel committed:
Work around infitite loop

thgoebel committed:
Add CT insertion and Security_Non_Discreditability_CT lemma

thgoebel committed:
Server: sign <epoch_id, chainhash> + put it into CT + let auditors verify signs

This fixes the attack on Non-Discreditability by a malicious CA

thgoebel committed:
Add VerifyEpoch to querying

felixlinker commented on tamarin/main.spthy:
```diff
@@ -0,0 +1,123 @@
+
+rule Init_Server:
+    let epoch_id = %1
+        latest_epoch_id = epoch_id
```
Same as `epoch_id`. Intended?

thgoebel replied:
Yes, for clarity.

Also for clarity I try to pull out things out into `let` assignments, instead of not having variables and setting the values directly in the state fact. To have explicit naming.

felixlinker commented on tamarin/main.spthy:
```diff
@@ -0,0 +1,123 @@
+    [ Fr(~id) ] --> [ St_Client(~id, $email, keys, keys_rev) ]
+
+rule Register_Server_Pk:
+    [ Fr(~sk) ] --> [ !Ltk($server, ~sk), !Pk($server, pk(~sk)), Out(pk(~sk)) ]
```
Maybe it would be clearer to rename `!Ltk` to `!LtkServer` to avoid confusion with user keys.

felixlinker commented on tamarin/publishepoch.spthy:
```diff
@@ -0,0 +1,188 @@
+      , RequestInsertion($email, keys_new)
+    ]->
+    [ St_Client(id, $email, keys_new, keys_rev_new)
+    , Out(<'insert', $email, ~newkey>) ]
```
I find it odd to out the key itself. I see why it might make no difference to out the public key instead, because we're not interested in protocol-level security features anyways, but I think the model would be clearer (and more future-proof) if you used `Out(pk(~newkey))`.

felixlinker commented on tamarin/publishepoch.spthy:
```diff
@@ -0,0 +1,188 @@
+// TODO: distinguish between the normal keylist and the not-yet-included keylist?
+// TODO: pk/sk
+rule ClientKeyGen:
+    let keys_new = keys ++ ~newkey
```
I think this multiset will lead you into _many_ problems. I also don't see the immediate value of it. My guess is that you wanted to go for something like: "When I see revisions, it's the same as all of my keys." To model something like this, you indeed need to use something like a multiset (although a tuple would also do as you can keep the order).

I instead suggest you to solve this on the formula level. You could instead of writing a lemma like "verification always terminates" write "whenever I verify some key as part of my tree, I generated this key." I suspect this will lead to much better model performance, and you also implicitly model that clients can seize to use keys.

felixlinker commented on tamarin/publishepoch.spthy:
```diff
@@ -0,0 +1,188 @@
+rule ReceiveInsertAndIssueEpoch[color=#6d4af6]:
+    let epoch_id_new = %latest_epoch_id %+ %1
+        roothash_new = 'todo'
+        chainhash_new = h(chainhash_old ++ roothash_new)
```
I would encourage you to use tuples here. Multisets don't preserve orders, and I expect very weird side effects from using multisets to model concatenation.

Additionally, you loose all benefits of the persistent fact approach if you actually compute hashes. What I explained to you in the meeting (using persistent facts), means that you simply include some element `x` in the logs, when you do something like this: `!InLog(x)`. Or do you go for the full model immediately?

_If_ you go for the full model immediately, I'd suggest that you follow the approach of a paper that I pointed out to you earlier: https://arxiv.org/abs/2303.04500

felixlinker commented on tamarin/publishepoch.spthy:
```diff
@@ -0,0 +1,188 @@
+    [ St_Client_VerifyEpoch0(id, email, keys, keys_rev, %epoch_id_to_verify, expected_root_hash, stack) ]
+    -->
+    [ St_Client_VerifyEpoch1(id, email, keys, keys_rev, %epoch_id_to_verify, expected_root_hash, stack)
+    , Out(<'ve1', %epoch_id_to_verify>) ]
```
Why do you use `Out` here to trigger the verification?

thgoebel replied:
To closely model reality, i.e. the API request to https://mail.proton.me/api/kt/v1/epochs/300

`rule VerifyEpoch_2` is executed by the server. The state fact `St_Server` is commented out because I didn't want to deal with that looping problem. The idea is that `!SST_` should only appear in server rules (because only the server is supposed to have access to **S**erver **ST**ate).

felixlinker commented on tamarin/publishepoch.spthy:
```diff
@@ -0,0 +1,188 @@
+    ]//, St_Server(id, sk, latest_epoch_id) ]
+    -->
+    //[ St_Server(id, sk, latest_epoch_id)
+    [ Out(<'ve2', chainhash_prev, chainhash_verify, roothash_verify>) ]
```
Again. Why `Out`?

thgoebel committed:
Rename !Ltk to !LtkServer (for clarity)

thgoebel committed:
Use tuple instead of multiset for chainhash

thgoebel committed:
Make pk() explicit

thgoebel committed:
Model tree inclusion proof with persistent fact

thgoebel committed:
Add (simplified/limited) Query Consistency security lemma

thgoebel committed:
WIP: Add revisions to allow key updates

WIP because proving QueryConsistency blows up and needs work

thgoebel committed:
refactor: Move ProtonVKD.VerifyEpoch into its own file

thgoebel committed:
WIP: try putting minEpochId into the !TreeLeaf

thgoebel committed:
WIP: implement tree-as-list (only for VKD.Query so far)

thgoebel committed:
Fix executability lemmas for tree-as-list

thgoebel committed:
refactor: Store key_revs and keys as pairs in St_Client

thgoebel committed:
Model SelfAudit

thgoebel committed:
Document tree model

thgoebel committed:
Big refactor

1. Querying doesn't need client-state
   => use a dedicated St_Query instead of sharing St_Client

2. VerifyEpoch: instead of a call_stack, emit a state fact and pick it up again.
   Here the ~id is important. It ensures correct rule progression.

thgoebel committed:
Use same colours for Query and VerifyQuery

thgoebel committed:
refactor: merge rules to simplify Query/VerifyQuery

thgoebel committed:
Implement incremental/delta SelfAudit

Instead of always checking the entire history,
only self-audit the revisions that have been added.

thgoebel committed:
Implement Apppend-only-ness Auditor

felixlinker commented:
I keep getting pinged by pushes (I know I can unsubscribe and will). But maybe worth closing anyways :)
